### PR TITLE
PHPLIB-1343 Add tests on Array Expression Operators

### PR DESCRIPTION
**Pull request by GromNaN**
Fix [PHPLIB-1343](https://jira.mongodb.org/browse/PHPLIB-1343)

- `$arrayElemAt`
- `$arrayToObject`
- `$concatArrays`
- `$filter`
- ~`$firstN`~ (already done)
- `$in`
- `$indexOfArray`
- `$isArray`
- ~`$lastN`~ (already done)
- `$map`
- `$maxN`
- `$minN`
- `$objectToArray` (skipped duplicate example with `$arrayToObject`)
- `$range`
- `$reduce`
- `$reverseArray`
- `$size`
- `$slice`
- `$sortArray`
- ~`$zip`~ (already done)

### CODE REVIEW

File: generator/config/expression/isArray.yaml [hunk 1]
```diff
@@ -3,12 +3,31 @@ name: $isArray
 link: 'https://www.mongodb.com/docs/manual/reference/operator/aggregation/isArray/'
 type:
     - resolvesToBool
-encode: array
+encode: single
```
Comment (GromNaN) [inline generator/config/expression/isArray.yaml]: The syntax [in the doc](https://www.mongodb.com/docs/manual/reference/operator/aggregation/isArray/) is confusing. It looks like the expression must be encapsulated in a single-value array, but in the example, the variable is used directly.
Maybe a doc issue.

File: tests/Builder/Expression/Pipelines.php [hunk 2]
```diff
@@ -171,41 +490,705 @@ enum Pipelines: string
...
+                    "$map": {
+                        "input": "$tempsC",
+                        "as": "tempInCelsius",
+                        "in": {
+                            "$add": [
+                                {
+                                    "$multiply": [
+                                        "$$tempInCelsius",
+                                        {
+                                            "$numberDouble": "1.8000000000000000444"
```
Comment (GromNaN) [inline tests/Builder/Expression/Pipelines.php:569]: Other example of numeric value incorrectly encoded.
Related to https://github.com/mongodb/mongo-php-builder/pull/21#discussion_r1448851990

File: src/Builder/BuilderEncoder.php [hunk 3]
```diff
@@ -278,6 +278,8 @@ private function recursiveEncode(mixed $value): mixed
             foreach (get_object_vars($value) as $key => $val) {
                 $value->{$key} = $this->recursiveEncode($val);
             }
+
+            return $value;
```
Comment (GromNaN) [inline src/Builder/BuilderEncoder.php:282]: Not a bug, but an optimization.

File: generator/config/expression/isArray.yaml (hunk 1 above)
Comment (alcaeus) [reply generator/config/expression/isArray.yaml]: I believe this is covered by this note in the docs:
> Aggregation expressions accept a variable number of arguments. These arguments are normally passed as an array. However, when the argument is a single value, you can simplify your code by passing the argument directly without wrapping it in an array.

I'd have to do some tests on how values are handled. Looking at the examples in the [Behavior](https://www.mongodb.com/docs/manual/reference/operator/aggregation/isArray/#behavior) section, we should make sure that passing a value of `[1]` is actually recognised as an array, which might require changing `encode` back to `array` and updating the examples in the config files.

**Review by alcaeus — COMMENTED** [review]
One suggestion regarding argument encoding for `$isArray`, but LGTM other than that.

File: generator/config/expression/isArray.yaml (hunk 1 above)
Comment (GromNaN) [reply generator/config/expression/isArray.yaml]: Ok, reverted to always use the array syntax.

**Review by alcaeus — APPROVED** [review]

**Pull request merged**